pkg/controller/controller.go

```diff
@@ -678,3 +735,34 @@ const (
 	reconcileDelete ReconciliationAction = "Delete"
 	reconcilePoll   ReconciliationAction = "Poll"
 )
+
+func (c *controller) getClusterID() (id string) {
+	// Use the read lock for reading, so that multiple instances
+	// provisioning at the same time do not collide.
+	c.clusterIDLock.RLock()
+	id = c.clusterID
+	c.clusterIDLock.RUnlock()
+
+	// fast exit if it exists
+	if id != "" {
+		return
+	}
+	// lazily create on first access if does not exist
+	c.clusterIDLock.Lock()
+	// make sure nobody set ID while we were locking
+	id = c.clusterID
+	if id != "" {
+		c.clusterIDLock.Unlock()
```

Comment:
could you just do `defer c.clusterIDLock.Unlock()` after you set the lock?

Reply:
you could also simplify it down to:
```
c.clusterIDLock.Lock()
if id = c.clusterID; id == "" {
    id = string(uuid.NewUUID())
    c.clusterID = id
}
c.clusterIDLock.Unlock()
```

Reply:
adjusted.